aten/src/ATen/native/TensorShape.cpp

```diff
@@ -392,6 +390,90 @@ Tensor cat(TensorList tensors, int64_t dim) {
   return result;
 }
 
+Tensor block_diag(TensorList tensors) {
+  // ScalarType
+  Tensor result;
+  if (tensors.size() == 0) {
+    result = at::empty({});
+  } else if (tensors.size() == 1) {
+    result = at::empty({}, tensors[0].options());
+  } else {
+    // Use at::result_type to find the right tensor to copy options from, so
+    // that the output data type is correct
+    const Tensor* options_tensor = &tensors[0];
+    ScalarType output_scalar_type = options_tensor->scalar_type();
+
+    for (size_t tensor_idx = 1; tensor_idx < tensors.size(); tensor_idx++) {
+      const Tensor* other_tensor = &tensors[tensor_idx];
+      ScalarType scalar_type = at::result_type(*options_tensor, *other_tensor);
+
+      if (scalar_type != output_scalar_type) {
+        output_scalar_type = scalar_type;
+        options_tensor = other_tensor;
+      }
+    }
+    result = at::empty({}, options_tensor->options());
+  }
+  return at::block_diag_out(result, tensors);
+}
+
+Tensor& block_diag_out(Tensor& result, TensorList tensors) {
```

Comment:
I added checks to make sure all the input tensors are on the same device. I shouldn't need to check the result's device since I'm copying the tensor options from one of the inputs.